velox/functions/prestosql/JsonFunctions.cpp

```diff
@@ -14,11 +14,92 @@
  * limitations under the License.
  */
 #include "velox/expression/VectorFunction.h"
+#include "velox/functions/prestosql/json/JsonStringUtil.h"
 #include "velox/functions/prestosql/json/SIMDJsonUtil.h"
 #include "velox/functions/prestosql/types/JsonType.h"
+#include "velox/type/Conversions.h"
 
 namespace facebook::velox::functions {
 
+StringView trimAndEscape(const char* data, size_t length) {
+  return velox::util::trimWhiteSpace(data, length);
+}
+
+namespace {
+constexpr const char* kArrayStart = "[";
+constexpr const char* kArrayEnd = "]";
+constexpr const char* kSeparator = ",";
+constexpr const char* kObjectStart = "{";
+constexpr const char* kObjectEnd = "}";
+constexpr const char* kObjectKeySeparator = ":";
+constexpr const char* kQuote = "\"";
+
+class JsonView {
+ public:
+  virtual void canonicalize(std::stringstream& stream) = 0;
```

Comment:
virtual dispatch will kill the performance here.  I think it should be possible to do the canonicalization with a linear scan (maybe with the help of a optional stack). There is no need to build a full tree for it.

Reply:
I will refactor it.